File: server/mc/1.21.1-neoforge/src/main/java/calebxzhou/rdi/mc/server/network/RServerNetwork.java
package calebxzhou.rdi.mc.server.network;

import calebxzhou.rdi.mc.common.RGlobalPlayerList;
import calebxzhou.rdi.mc.server.firmsection.FirmSectionService;
import com.google.gson.Gson;
import net.minecraft.server.MinecraftServer;
import net.minecraft.server.dedicated.DedicatedServer;
import net.minecraft.server.level.ServerPlayer;
import net.neoforged.bus.api.SubscribeEvent;
import net.neoforged.fml.common.EventBusSubscriber;
import net.neoforged.neoforge.network.PacketDistributor;
import net.neoforged.neoforge.network.event.RegisterPayloadHandlersEvent;

import java.util.ArrayList;

@EventBusSubscriber(modid = "rdi")
public final class RServerNetwork {
    private static final Gson GSON = new Gson();
    private static volatile RdiGlobalPlayerListPayload lastPayload;

    private RServerNetwork() {
    }

    @SubscribeEvent
    public static void registerPayloads(RegisterPayloadHandlersEvent event) {
        event.registrar("1")
                .optional()
                .playToClient(RdiGlobalPlayerListPayload.TYPE, RdiGlobalPlayerListPayload.STREAM_CODEC, (payload, context) -> {
                })
                .playToClient(RFirmSectionsPayload.TYPE, RFirmSectionsPayload.STREAM_CODEC, (payload, context) -> {
                });
    }

    public static void sendToAll(DedicatedServer server, RGlobalPlayerList playerList) {
        var payload = new RdiGlobalPlayerListPayload(GSON.toJson(playerList));
        lastPayload = payload;
        for (var player : server.getPlayerList().getPlayers()) {
            PacketDistributor.sendToPlayer(player, payload);
        }
    }

    public static void sendLastTo(net.minecraft.server.level.ServerPlayer player) {
        var payload = lastPayload;
        if (payload != null) {
            PacketDistributor.sendToPlayer(player, payload);
        }
    }

    public static void sendFirmSectionsToAll(MinecraftServer server) {
        var payload = firmSectionsPayload(server);
        for (var player : server.getPlayerList().getPlayers()) {
            PacketDistributor.sendToPlayer(player, payload);
        }
    }

    public static void sendFirmSectionsTo(ServerPlayer player) {
        PacketDistributor.sendToPlayer(player, firmSectionsPayload(player.server));
    }

    private static RFirmSectionsPayload firmSectionsPayload(MinecraftServer server) {
        var entries = new ArrayList<RFirmSectionsPayload.Entry>();
        for (var section : FirmSectionService.INSTANCE.all(server)) {
            entries.add(new RFirmSectionsPayload.Entry(
                    section.getDimensionId(),
                    section.getChunkX(),
                    section.getSectionY(),
                    section.getChunkZ()
            ));
        }
        return new RFirmSectionsPayload(entries);
    }
}

File: server/mc/1.21.1-neoforge/src/main/kotlin/calebxzhou/rdi/mc/server/network/RServerNetwork.kt
package calebxzhou.rdi.mc.server.network

import calebxzhou.rdi.mc.common.RGlobalPlayerList
import calebxzhou.rdi.mc.server.firmsection.FirmSectionService.all
import com.google.gson.Gson
import net.minecraft.server.MinecraftServer
import net.minecraft.server.dedicated.DedicatedServer
import net.minecraft.server.level.ServerPlayer
import net.neoforged.bus.api.SubscribeEvent
import net.neoforged.fml.common.EventBusSubscriber
import net.neoforged.neoforge.network.PacketDistributor
import net.neoforged.neoforge.network.event.RegisterPayloadHandlersEvent
import net.neoforged.neoforge.network.handling.IPayloadContext
import net.neoforged.neoforge.network.handling.IPayloadHandler
import kotlin.concurrent.Volatile

@EventBusSubscriber(modid = "rdi")
object RServerNetwork {
    private val GSON = Gson()

    @Volatile
    private var lastPayload: RGlobalPlayerListPayload? = null
@JvmStatic
    @SubscribeEvent
    fun registerPayloads(event: RegisterPayloadHandlersEvent) {
        event.registrar("1")
            .optional()
            .playToClient(
                RGlobalPlayerListPayload.TYPE,
                RGlobalPlayerListPayload.STREAM_CODEC,
                IPayloadHandler { _: RGlobalPlayerListPayload, _: IPayloadContext -> })
            .playToClient(
                RFirmSectionsPayload.TYPE,
                RFirmSectionsPayload.STREAM_CODEC,
                IPayloadHandler { _: RFirmSectionsPayload, _: IPayloadContext -> })
    }

    fun sendToAll(server: DedicatedServer, playerList: RGlobalPlayerList?) {
        val payload = RGlobalPlayerListPayload(GSON.toJson(playerList))
        lastPayload = payload
        for (player in server.getPlayerList().getPlayers()) {
            PacketDistributor.sendToPlayer(player, payload)
        }
    }

    fun sendLastTo(player: ServerPlayer) {
        lastPayload?.let { PacketDistributor.sendToPlayer(player, it) }
    }

    fun sendFirmSectionsToAll(server: MinecraftServer) {
        val payload = firmSectionsPayload(server)
        for (player in server.getPlayerList().getPlayers()) {
            PacketDistributor.sendToPlayer(player, payload)
        }
    }

    fun sendFirmSectionsTo(player: ServerPlayer) {
        PacketDistributor.sendToPlayer(player, firmSectionsPayload(player.server))
    }

    private fun firmSectionsPayload(server: MinecraftServer): RFirmSectionsPayload {
        val entries = all(server).map { section ->
            RFirmSectionsPayload.Entry(
                section.dimensionId,
                section.chunkX,
                section.sectionY,
                section.chunkZ
            )
        }
        return RFirmSectionsPayload(entries)
    }
}
